### ISSUE TITLE
why not using set_requires_all_reduce for hsdp

### ISSUE THREAD

**Issue by samsja** [pytorch/torchtitan#1929]
It seems that the hsdp (using dp_replicate) and the grad accumulation logic is not optimal.

It seems that the global all reduce is triggered on each grad accumulation step whereas it should only be triggered in the last one. Adding model.set_requires_all_reduce(micro_step == len(micro_batches) - 1)
https://docs.pytorch.org/docs/stable/distributed.fsdp.fully_shard.html#torch.distributed.fsdp.FSDPModule.set_requires_all_reduce would allow to trigger the all reduce only in the last step.

My guess is that in practice nobody is doing more than one grad accumulation when using the dp_replicate, but still would be nice to add to be feature complete imo

**Comment by tianyu-l** [comment]
oh, good point.

For pure FSDP, there's a tradeoff of whether to perform reduce scatter only once or every grad accumulation step.
But for HSDP, it seems there's no benefit of performing the all-reduce every step.

But like you said 
> in practice nobody is doing more than one grad accumulation when using the dp_replicate

The point of grad accumulation is to save memory to fit in training (with smaller batch per step and more steps, hence slower);
the point of HSDP is to sacrifice memory to trade for throughput. They are quite the opposite?

If we could argue that no one should ever do grad accumulation + HSDP, then the optimization might be a fake problem.

**Comment by samsja** [comment]
> The point of grad accumulation is to save memory to fit in training (with smaller batch per step and more steps, hence slower);
the point of HSDP is to sacrifice memory to trade for throughput. They are quite the opposite?

I think there is a bit more to this story. HSDP with large batch size can also help with cluster with bad interconnect across nodes. Also people might be using torchtitan to run research ablation around batch size in which case they might want to artificially increase batch size. Also when doing RL (if someone use torchtitan as backend) you have very large batch size.

My take is that its worth adding as its a one line change that doesn't do anything when there is only one grad acc in most usual setup but help in other. Also I believe people look at torchtitan for having the best reference on how to use fsdp2 so defo nice to showcase all the feature :)

I am happy to draft a PR if this sounds good to you.

**Comment by tianyu-l** [comment]
> HSDP with large batch size can also help with cluster with bad interconnect across nodes.

This makes sense. What's related is that HSDP can be used in torchft fault-tolerant training.

> Also people might be using torchtitan to run research ablation around batch size in which case they might want to artificially increase batch size. Also when doing RL (if someone use torchtitan as backend) you have very large batch size.

For these two, why can't they just use FSDP?

> I am happy to draft a PR if this sounds good to you.

Sure, please. Thank you!

**Comment by EquationWalker** [comment]
By the way, unlike `set_requires_gradient_sync`, `set_requires_all_reduce` does not incur an additional memory burden.
@tianyu-l

**Comment by rakkit** [comment]
i just saw this issue today, 

> in practice nobody is doing more than one grad accumulation when using the dp_replicate

this is not very true.   in practice, especially for people who want to train model over few T tokens, the optimal BS is really large.   in that sense the only way to do it is HSDP.    and people may have to run HSDP on total of 1000 gpus or 2000 gpus which can be 1) not possible to  have that much of gpus 2) need pay for scaling. So HSDP on few hundred gpus + grad accumulation is the most realistic way.

**Comment by tianyu-l** [comment]
@samsja would you still like to prepare a PR?

**Comment by garrett361** [comment]
Also just noticed this today. (Hi @rakkit! We're always on the same issues lol). I was a little surprised this wasn't already implemented.

> I think there is a bit more to this story. HSDP with large batch size can also help with cluster with bad interconnect across nodes. Also people might be using torchtitan to run research ablation around batch size in which case they might want to artificially increase batch size. Also when doing RL (if someone use torchtitan as backend) you have very large batch size.

Agree with all of this @samsja. Another example: HSDP + grad accum can also be very helpful when your batches have highly variable compute costs, e.g. packed data with widely varying sequence lengths and hence varying attention costs and each FSDP replica is also a CP group, say. The entire world is bottlenecked by the time it takes most expensive/slowest batch to finish, and having a larger batch size/grad accum rate helps homogenize costs across replicas

**Comment by samsja** [comment]
> Also just noticed this today. (Hi [@rakkit](https://github.com/rakkit)! We're always on the same issues lol). I was a little surprised this wasn't already implemented.
> 
> > I think there is a bit more to this story. HSDP with large batch size can also help with cluster with bad interconnect across nodes. Also people might be using torchtitan to run research ablation around batch size in which case they might want to artificially increase batch size. Also when doing RL (if someone use torchtitan as backend) you have very large batch size.
> 
> Agree with all of this [@samsja](https://github.com/samsja). Another example: HSDP + grad accum can also be very helpful when your batches have highly variable compute costs, e.g. packed data with widely varying sequence lengths and hence varying attention costs and each FSDP replica is also a CP group, say. The entire world is bottlenecked by the time it takes most expensive/slowest batch to finish, and having a larger batch size/grad accum rate helps homogenize costs across replicas

100% agree . Ah damn let me just do a pr quickly haha

**Comment by garrett361** [comment]
Yeah sorry didn't have time to put one up right now. Can try to do it this week if you can't, or can help review and test if you get there first

**Comment by tianyu-l** [comment]
@garrett361 
> The entire world is bottlenecked by the time it takes most expensive/slowest batch to finish, and having a larger batch size/grad accum rate helps homogenize costs across replicas

For this particular point, could you explain a bit more? I expect something like https://arxiv.org/abs/2601.19362 could help, but if FSDP/HSDP still synchronizes per microbatch, how could grad accumulation alleviate load imbalance?

**Comment by garrett361** [comment]
@tianyu-l the setup I have in mind is where your minibatches have variable compute costs, e.g. due to packing and differing seqlens among the packed examples. So, each minibatch might have the same computational cost in expectation, but the variance is non-trivial.

In that scenario, all ranks need to wait for the slow outlier at every all-reduce point, and it can be a real efficiency drag. The worst cases are FSDP or HSDP with all-reduces every step, where you have to wait for the slowest minibatch for every fwd/bwd. If you're using gradient accumulation, then you can mitigate some of this by using HSDP and only doing the global all-reduce once every `accum_steps` iterations, which is not only fewer collectives, but also helps reduce the variance across the compute times on different GPUs.

You can see this in an idealized limit where the microbatch costs are very bi-modal, with typical compute taking either `t_long` or `t_short` seconds, say. Let's make it even more idealized and say every DP group flips between one case and the other every microbatch, with at least one `t_long` in every fwd. Then, if you're using FSDP or HSDP without the `set_requires_all_reduce` optimization, compute time is at least `~ n_microbatches * t_long` due to all the sync points. If you use HSDP with `set_requires_all_reduce`, then you only sync at the final all-reduce and the floor is instead set by the average cost across microbatches, instead of the sum of the wost-cases: `~ n_microbatches * (t_long + t_short)/2`. A super-not-realistic scenario, but versions of this hold in more realistic ones.

[I mocked this up here](https://gist.github.com/garrett361/8cf4f612eec730fe66881a21fd9449d2). Simple all-FFN model where variable compute costs are enforced by changing the batch size of the different microbatches. Half the GPUs are given `batch_size=1` and the other half `batch_size=2`, alternating assignments every microbatch. E2E tps improvement from delaying the all-reduce is ~25% 

```
❯ torchrun --nproc-per-node=4 scripts/bench_delay_all_reduce.py --accum-steps 2 --batch-size-ratio 2 --delay-all-reduce
HSDP(delayed AR) | accum_steps: 2, batch_size_ratio: 2 | step: 2145.8 ± 2.5 ms | tok/sec/gpu: 5727 |

❯ torchrun --nproc-per-node=4 scripts/bench_delay_all_reduce.py --accum-steps 2 --batch-size-ratio 2
HSDP | accum_steps: 2, batch_size_ratio: 2 | step: 2698.9 ± 1.2 ms | tok/sec/gpu: 4553 |

❯ torchrun --nproc-per-node=4 scripts/bench_delay_all_reduce.py --accum-steps 2 --batch-size-ratio 2 --fsdp-only
FSDP | accum_steps: 2, batch_size_ratio: 2 | step: 2693.5 ± 1.6 ms | tok/sec/gpu: 4562 |
```
The simple analysis above says HSDP w/ delayed AR should be faster by a factor of about `r = 2 * t_long / (t_short + t_long)` and here `t_long ~= 2 * t_short` implying r = 4/3 or ~33% speedup. So, seems roughly realized.

@samsja @rakkit similar mental model?

**Comment by rakkit** [comment]
I don’t really remember why I replied to this thread at the beginning, but here are a few things we observed:

1. In theory, the optimal batch size scales with the square root of the number of training tokens. So the more data we want to train on, the larger the batch size needs to be to stay optimal. This is regardless of model size, which means we do need a very large batch size.
2. FSDP starts to scale poorly after a certain point, so we need HSDP, especially when using optimizers like Muon.
3. Sometimes people simply cannot get enough GPUs to fit that giant batch size.
4. I think the discussion was about F/HSDP × gradient accumulation. Based on what we observed, there is a trade-off: we can reserve some memory and delay reduce, effectively making it “all-reduce only on the last accumulation step.”

From what we saw, we can at least apply this delayed reduce to the embedding, the first and last 1–2 decoder layers, and the linear head. These 4–6 blocks are the most critical parts because they usually cannot be overlapped well, and matters for gradient accumulation

* At the beginning of each forward pass, we need to wait for the embedding and first layer to be gathered.
* At the end of each backward pass, we need to wait for the first layer and embedding reduce to finish before starting `optimizer.step()`.
* At the beginning of each backward pass, depending on the setup, the output linear gradient reduce can sometimes block the all-gather of the second-to-last layer. This happens when the output linear reduce is slower than the last decoder layer’s backward pass, for example with a 200k vocab and small models, such as <30B MoE.

**Comment by rakkit** [comment]
and for CP, yes it is even worse.... even without gradient accumulation, CP+HSDP kind naturally broken the compute-comm badly.   this can get even much much worse when we training hybrid model, e.g. 3:1 linear/swa attention + full attention.  in CP the full attention is always slow and it blocked all rank's gather.

**Comment by samsja** [comment]
i can't believe I didn't took the time to do a pr in the era of agent, my bad thanks for fixing it
